lighthouse-core/audits/dobetterweb/uses-http2.js

```diff
@@ -43,67 +58,174 @@ class UsesHTTP2Audit extends Audit {
     return {
       id: 'uses-http2',
       title: str_(UIStrings.title),
-      failureTitle: str_(UIStrings.failureTitle),
       description: str_(UIStrings.description),
-      requiredArtifacts: ['URL', 'devtoolsLogs'],
+      requiredArtifacts: ['URL', 'devtoolsLogs', 'traces'],
     };
   }
 
+  /**
+   * Computes the estimated effect all results being converted to use http/2, the max of:
+   *
+   * - end time of the last long task in the provided graph
+   * - end time of the last node in the graph
+   *
+   * @param {Array<{url: string}>} results
+   * @param {Node} graph
+   * @param {Simulator} simulator
+   * @return {number}
+   */
+  static computeWasteWithTTIGraph(results, graph, simulator) {
+    const beforeLabel = `uses-http2-before`;
+    const afterLabel = `uses-http2-after`;
+    const flexibleOrdering = true;
+
+    const urlsToChange = new Set(results.map(result => result.url));
+    const simulationBefore = simulator.simulate(graph, {label: beforeLabel, flexibleOrdering});
+
+    // Update all the protocols to reflect implementing our recommendations
+    /** @type {Map<string, string>} */
+    const originalProtocols = new Map();
+    graph.traverse(node => {
+      if (node.type !== 'network') return;
+      if (!urlsToChange.has(node.record.url)) return;
+
+      originalProtocols.set(node.record.requestId, node.record.protocol);
+      node.record.protocol = 'h2';
+    });
+
+    const simulationAfter = simulator.simulate(graph, {label: afterLabel, flexibleOrdering});
+
+    // Restore the original protocol after we've done our simulation
+    graph.traverse(node => {
+      if (node.type !== 'network') return;
+      const originalProtocol = originalProtocols.get(node.record.requestId);
+      if (originalProtocol === undefined) return;
+      node.record.protocol = originalProtocol;
+    });
+
+    const savingsOnOverallLoad = simulationBefore.timeInMs - simulationAfter.timeInMs;
+    const savingsOnTTI = Interactive.getLastLongTaskEndTime(simulationBefore.nodeTimings) -
+      Interactive.getLastLongTaskEndTime(simulationAfter.nodeTimings);
+    const savings = Math.max(savingsOnTTI, savingsOnOverallLoad);
+
+    // Round waste to nearest 10ms
+    return Math.round(Math.max(savings, 0) / 10) * 10;
+  }
+
+  /**
+   * Determines whether a network request is a "static resource" that would benefit from H2 multiplexing.
+   * XHRs, tracking pixels, etc generally don't benefit as much because they aren't requested en-masse
+   * for the same origin at the exact same time.
+   *
+   * @param {LH.Artifacts.NetworkRequest} networkRequest
+   * @return {boolean}
+   */
+  static isStaticAsset(networkRequest) {
+    if (!STATIC_RESOURCE_TYPES.has(networkRequest.resourceType)) return false;
```

Comment:
I don't like the assumptions being made here. I can totally see a site making a lot of important XHR calls at the same time.

Perhaps the `length > 6` check could instead be expanded to be `length > 6 for some window of time`?

What happens if all resources are changed to `h2` in the simulation? It seems you're doing this because even if there were just 3 h1 requests to an origin, Lantern would suppose things were faster than they'd really be, but I'm missing the "why" here.

Reply:
That's true, XHRs could be requested with more than 6 at a time. The real reason is a combination of "XHRs, tracking pixels, etc generally don't benefit as much because they aren't requested en-masse for the same origin at the exact same time." and simulation limitations "The simulator approximation of HTTP/2 is a little more generous than reality.". XHRs are a place where if they *aren't* requested en-masse then we will absolutely overpromise on the savings. Given that it's fairly rare for >6 XHRs to be inflight at once on an origin that didn't also serve at least 6 static assets. The tradeoff seems reasonable.

Happy to add whatever you comment you write summarizing this that makes it clearer :)